### PR TITLE
Fixed #27452 - Added SerialField.

### PR DESCRIPTION
**Pull request by csirmazbendeguz**
# Trac ticket number
ticket-27452

# Branch description
A basic implementation of serial fields without support for altering field types.

It adds SerialFields without the complexity of supporting altering between different types (see previous PR).

[Previous PR](https://github.com/django/django/pull/18123)

# Checklist
- [X] This PR targets the `main` branch. 
- [X] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [X] I have checked the "Has patch" ticket flag in the Trac system.
- [X] I have added or updated relevant tests.
- [X] I have added or updated relevant docs, including release notes if applicable.
- [X] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

**Review by LilyFoote — APPROVED** [review]
Looks good to me!

File: docs/ref/contrib/postgres/fields.txt
Comment (charettes) [inline docs/ref/contrib/postgres/fields.txt:933]: Maybe we should mention that it's backed by the Postgres `serial` type?

**Review by charettes — COMMENTED** [review]
I think we need to override `rel_db_type` as well. Do we have any tests that covers foreign keys to a unique `SerialField`?

**Comment by charettes** [comment]
Apart from the `get_prep_value` question and the possible issues with foreign keys to a unique `SerialField` attempted to be created as serials this is looking great to me!

File: docs/ref/contrib/postgres/fields.txt
Comment (csirmazbendeguz) [reply docs/ref/contrib/postgres/fields.txt:933]: ok

File: django/contrib/postgres/fields/serial.py
Comment (csirmazbendeguz) [reply django/contrib/postgres/fields/serial.py:37]: It's to allow `SerialModel.objects.create(serial=None)`. If we want to disable this, it can be removed.

The reason it's not in `AutoFieldMixin` is because primary keys are a special case, if it's set to `None`, it's omitted from the `INSERT` query.

**Comment by csirmazbendeguz** [comment]
Thanks a lot @charettes , you were right ! I wasn't aware. I pushed the foreign key fix.

**Comment by LilyFoote** [comment]
Looking at the differences between this PR and the other, I'm wondering if we should move the `inspectdb` changes into this one too. If I understand correctly, they're independent of the migration story that's causing difficulty.

**Comment by csirmazbendeguz** [comment]
> Looking at the differences between this PR and the other, I'm wondering if we should move the `inspectdb` changes into this one too. If I understand correctly, they're independent of the migration story that's causing difficulty.

@LilyFoote , I haven't added the introspection changes here because I tried to keep this PR minimal. If you agree, I would open a separate PR to add that once this is merged. It could be added separately, especially because I feel like the smaller I keep my PRs the faster they can be reviewed and merged by the maintainers.

It's also questionable if we want to add that at all, since we don't have introspection for any other Postgres field either. The only reason I added it to the original PR was that I was using introspection in the tests.

File: docs/releases/5.2.txt
Comment (sarahboyce) [inline docs/releases/5.2.txt:76]: @nessita do you think we should promote this a headline feature?

(nit: we will change this to present tense but want to check where we should put it first)

File: tests/postgres_tests/test_serial.py
Comment (sarahboyce) [inline tests/postgres_tests/test_serial.py:127]: I think we should either break this test up into smaller cases with descriptive test names for what we're checking or add one or two comments. I was trying to see if any of the object creation is unnessary

This is maybe a suggestion, we can probably do better
```python
        # When setting to None, will continue the sequence.
        obj_3 = SerialModel.objects.create(
            small_serial=None,
            serial=None,
            big_serial=None,
        )
        self.assertEqual(obj_3.small_serial, 3)
        self.assertEqual(obj_3.serial, 3)
        self.assertEqual(obj_3.big_serial, 3)

        # Serial fields do not need to be unique.
        manual_obj_1 = SerialModel.objects.create(
            small_serial=1,
            serial=1,
            big_serial=1,
        )
        self.assertEqual(manual_obj_1.small_serial, 1)
        self.assertEqual(manual_obj_1.serial, 1)
        self.assertEqual(manual_obj_1.big_serial, 1)

        # The sequence is not interrupted by other objects.
        manual_obj_5 = SerialModel.objects.create(
            small_serial=5,
            serial=5,
            big_serial=5,
        )
        self.assertEqual(manual_obj_5.small_serial, 5)
        self.assertEqual(manual_obj_5.serial, 5)
        self.assertEqual(manual_obj_5.big_serial, 5)

        obj_4 = SerialModel.objects.create()
        self.assertEqual(obj_4.small_serial, 4)
        self.assertEqual(obj_4.serial, 4)
        self.assertEqual(obj_4.big_serial, 4)

        obj_5 = SerialModel.objects.create()
        self.assertEqual(obj_5.small_serial, 5)
        self.assertEqual(obj_5.serial, 5)
        self.assertEqual(obj_5.big_serial, 5)
```

File: tests/postgres_tests/test_serial.py
Comment (sarahboyce) [inline tests/postgres_tests/test_serial.py:193]:
```suggestion
                self.assertIsNone(SerialModel(**kwargs).full_clean())
```
Optional

File: tests/postgres_tests/test_serial.py
Comment (sarahboyce) [inline tests/postgres_tests/test_serial.py:320]:
```suggestion
```
I think we do basically this inside `SerialFKModelTests.test_create` so we probably don't need this

File: docs/ref/contrib/postgres/fields.txt
Comment (sarahboyce) [inline docs/ref/contrib/postgres/fields.txt:937]:
```suggestion
    This uses the PostgreSQL ``serial`` type. See `the PostgreSQL documentation
    for the full details <https://www.postgresql.org/docs/current/
    datatype-numeric.html#DATATYPE-SERIAL>`__.
```

**Review by sarahboyce — COMMENTED** [review]
🥳 looks really good
I also checked loading fixtures and in the admin, all working nicely 👍

File: tests/postgres_tests/test_serial.py
Comment (csirmazbendeguz) [reply tests/postgres_tests/test_serial.py:193]: `full_clean` never returns anything so I'm not sure what you're suggesting?

File: tests/postgres_tests/test_serial.py
Comment (csirmazbendeguz) [reply tests/postgres_tests/test_serial.py:130]: I added some comments, let me know if that's acceptable?

**Comment by csirmazbendeguz** [comment]
Thanks a lot @sarahboyce ! I addressed your comments.

File: tests/postgres_tests/test_serial.py
Comment (sarahboyce) [reply tests/postgres_tests/test_serial.py:130]: Yes, I like it thank you 👍

File: tests/postgres_tests/test_serial.py
Comment (sarahboyce) [reply tests/postgres_tests/test_serial.py:190]: `.full_clean()` will either raise an error or return None, so in some of the existing tests around `.full_clean()` it has an `assertIsNone` so that the test has an assert
It's not ideal to have tests without asserts, another option I've seen is saving afterwards and asserting the values

Comment (csirmazbendeguz) [reply tests/postgres_tests/test_serial.py]: Okay, if it's a matter of consistency then of course let's do it this way, I just personally disagree.

File: docs/releases/5.2.txt
Comment (sarahboyce) [inline docs/releases/5.2.txt:76]:
```suggestion
* The new fields :class:`~django.contrib.postgres.fields.SerialField`,
  :class:`~django.contrib.postgres.fields.BigSerialField` and
  :class:`~django.contrib.postgres.fields.SmallSerialField` are automatically
  incrementing integer fields that use the PostgreSQL ``serial`` type. These
  are similar to :class:`~django.db.models.AutoField` except that they do not
  have to be a primary key, and are not limited to a single instance per model.
```

**Comment by sarahboyce** [comment]
(You are very welcome to add yourself to the [AUTHORS](https://github.com/django/django/blob/main/AUTHORS) if you wish to be listed 🥇)

File: docs/releases/5.2.txt
Comment (sarahboyce) [reply docs/releases/5.2.txt]: On reflection, let's leave it for now. I know you have bigger things planned and we can always promote it later 👍

**Comment by csirmazbendeguz** [comment]
Thanks @sarahboyce , I added all suggestions.

**Review by sarahboyce — APPROVED** [review]
Awesome work 🎊

File: django/db/backends/postgresql/operations.py
Comment (adamchainz) [inline django/db/backends/postgresql/operations.py:5]: Ideally, contrib apps would not be imported by core Django, nor require any knowledge within the core. They should act like third-party packages, with no privileged treatment...

**Comment by csirmazbendeguz** [comment]
[forum discussion](https://forum.djangoproject.com/t/serialfields-yay-or-nay/32541)

**Pull request closed without merging**
